### PR TITLE
MBS-11456 (SQL): Add MBIDs to artist credits

### PR DESCRIPTION
**Pull request by yvanzo**
Implement the SQL part for MBS-11456:

* Add `gid` column (for MBID) to `artist_credits` table
* Add replicated `artist_credit_gid_redirect` core table
* Write migration script
* Update scripts accordingly
* Update `artist_credit_gid_redirect` table when deleting artist credits.

### CODE REVIEW

File: admin/sql/CreateTables.sql [hunk 1]
```diff
@@ -376,13 +376,20 @@ CREATE TABLE artist_tag_raw
 
 CREATE TABLE artist_credit ( -- replicate
     id                  SERIAL,
+    gid                 UUID NOT NULL,
```
Comment (mwiencek) [inline admin/sql/CreateTables.sql:379]: This'll have to move to the end of the table definition

File: admin/sql/updates/schema-change/26.slave.sql [hunk 2]
```diff
@@ -409,4 +410,30 @@ BEGIN
...
+
+-- Generating GIDs
+UPDATE artist_credit SET gid =
+    generate_uuid_v3('6ba7b8119dad11d180b400c04fd430c8', 'artist_credit' || id);
+
+-- Adding NOT NULL constraint
+ALTER TABLE artist_credit ALTER COLUMN gid SET NOT NULL;
+
+-- Creating index
+CREATE UNIQUE INDEX artist_credit_idx_gid ON artist_credit (gid);
```
Comment (reosarevok) [inline admin/sql/updates/schema-change/26.slave.sql:430]: Shouldn't this also be added to CreateIndexes?

File: admin/sql/CreateTables.sql (hunk 1 above)
Comment (yvanzo) [reply admin/sql/CreateTables.sql]: Fixed at https://github.com/metabrainz/musicbrainz-server/pull/2092/commits/2ad5ae0c4b60d2fd7391bb48188e06a72454a26a#diff-92455c89deea6ff72051d26016d7124841420e89902f3257293b3f142f70ac1cR384

**Comment by reosarevok** [comment]
Changed the base, remember to also change it to schema 27 when fixing the conflicts :)

File: admin/sql/CreateReplicationTriggers.sql [hunk 3]
```diff
@@ -99,6 +99,10 @@ CREATE TRIGGER "reptg_artist_credit"
 AFTER INSERT OR DELETE OR UPDATE ON "artist_credit"
 FOR EACH ROW EXECUTE PROCEDURE "recordchange" ();
 
+CREATE TRIGGER "reptg_artist_credit_gid_redirect"
+AFTER INSERT OR DELETE OR UPDATE ON "artist_credit_gid_redirect"
+FOR EACH ROW EXECUTE PROCEDURE "recordchange" ('verbose');
+
```
Comment (reosarevok) [inline admin/sql/CreateReplicationTriggers.sql:105]: We'll also need a reptg2 trigger for dbmirror2 if/once we merge that (we should probably review the dbmirror2 triggers once we've done merging all schema change code).

File: lib/MusicBrainz/Server/Entity/ArtistCredit.pm [hunk 4]
```diff
@@ -17,6 +17,23 @@ with 'MusicBrainz::Server::Entity::Role::Editable';
...
+    isa => 'ArrayRef[Str]',
+    default => sub { [] },
+    traits => [ 'Array' ],
+    handles => {
+        add_gid_redirect => 'push',
+        clear_gid_redirects => 'clear',
+        all_gid_redirects => 'elements',
+    }
+);
+
```
Comment (reosarevok) [inline lib/MusicBrainz/Server/Entity/ArtistCredit.pm:36]: Not super important, but should this be a role used by both this and CoreEntity?

Comment (yvanzo) [reply lib/MusicBrainz/Server/Entity/ArtistCredit.pm]: Refactored in commit https://github.com/metabrainz/musicbrainz-server/pull/2092/commits/9d0f4d1a7e6bd4ee2cf9dced1eb9324373c4e1af.

**Review by reosarevok — APPROVED** [review]
Other than the one typo, this seems fine. I tested it a bit locally, making sure that new GIDs are created when creating a new artist and whatnot and that they're passed to the React layer already so we can use them whenever. I trust the tests for merging test that correctly.

File: t/lib/t/MusicBrainz/Server/Data/ArtistCredit.pm [hunk 5]
```diff
@@ -75,8 +75,10 @@ test 'Merging should combine ACs which are string-identical before merge (MBS-74
         SQL
 
     is($c->sql->select_single_value('SELECT count(*) FROM artist_credit'), 2, 'Two artist credits before merge');
+    is($c->sql->select_single_value('SELECT count(*) FROM artist_credit_gid_redirect'), 0, 'No artist credit MBID redirect before merge');
     $c->model('ArtistCredit')->merge_artists(2, [ 3 ]);
     is($c->sql->select_single_value('SELECT count(*) FROM artist_credit'), 1, 'One artist credit after merge');
+    is($c->sql->select_single_value('SELECT count(*) FROM artist_credit_gid_redirect'), 1, 'One artist credit MBID redirect before merge');
```
[the remark this answers is not in the document]
Comment (yvanzo) [reply t/lib/t/MusicBrainz/Server/Data/ArtistCredit.pm]: yes, updated commit https://github.com/metabrainz/musicbrainz-server/pull/2092/commits/a0a97874bc21623b007c6d72fe454b03acfdae26

**Pull request merged**